File: bindings/python/tests/test_mvbpp.py
"""
mvbpp Python binding tests.

Run with:  pytest bindings/python/tests/ -v
The mvbpp wheel must be installed first:
    pip install dist/mvbpp-*.whl
"""
import json
import os
import struct
import tempfile
from pathlib import Path

import pytest

import mvbpp

TESTS_DIR = Path(__file__).parent


# ── helpers ──────────────────────────────────────────────────────────────────

def load_magnetic_json(filename: str) -> str:
    """Return the 'magnetic' object from a test data file as a JSON string."""
    path = TESTS_DIR / filename
    data = json.loads(path.read_text())
    return json.dumps(data["magnetic"])


def is_step_bytes(data: bytes) -> bool:
    """Minimal check: STEP files start with 'ISO-10303-21'."""
    return data[:12] == b"ISO-10303-21"


# ── draw_magnetic_to_bytes ────────────────────────────────────────────────────

class TestDrawMagneticToBytes:

    def test_returns_bytes(self):
        j = load_magnetic_json("concentric_basic.json")
        result = mvbpp.draw_magnetic_to_bytes(j)
        assert isinstance(result, bytes)

    def test_returns_nonempty(self):
        j = load_magnetic_json("concentric_basic.json")
        result = mvbpp.draw_magnetic_to_bytes(j)
        assert len(result) > 1000

    def test_is_valid_step_header(self):
        j = load_magnetic_json("concentric_basic.json")
        result = mvbpp.draw_magnetic_to_bytes(j)
        assert is_step_bytes(result), "Output does not start with ISO-10303-21"

    def test_step_contains_geometry_data(self):
        j = load_magnetic_json("concentric_basic.json")
        result = mvbpp.draw_magnetic_to_bytes(j)
        text = result.decode("ascii", errors="replace")
        # STEP files with solid geometry contain CLOSED_SHELL or ADVANCED_BREP
        assert "CLOSED_SHELL" in text or "ADVANCED_BREP" in text

    def test_etd49_returns_valid_step(self):
        j = load_magnetic_json("ETD49_N87_10uH_5T.json")
        result = mvbpp.draw_magnetic_to_bytes(j)
        assert is_step_bytes(result)
        assert len(result) > 10_000

    def test_etd49_larger_than_basic(self):
        basic = mvbpp.draw_magnetic_to_bytes(load_magnetic_json("concentric_basic.json"))
        etd   = mvbpp.draw_magnetic_to_bytes(load_magnetic_json("ETD49_N87_10uH_5T.json"))
        # ETD49 with 5 turns should produce more geometry than the basic example
        assert len(etd) > len(basic)


# ── draw_magnetic (to file) ───────────────────────────────────────────────────

class TestDrawMagneticToFile:

    def test_writes_file(self):
        j = load_magnetic_json("concentric_basic.json")
        with tempfile.NamedTemporaryFile(suffix=".step", delete=False) as f:
            path = f.name
        try:
            mvbpp.draw_magnetic(j, path)
            assert os.path.exists(path)
            assert os.path.getsize(path) > 1000
        finally:
            os.unlink(path)

    def test_written_file_is_valid_step(self):
        j = load_magnetic_json("concentric_basic.json")
        with tempfile.NamedTemporaryFile(suffix=".step", delete=False) as f:
            path = f.name
        try:
            mvbpp.draw_magnetic(j, path)
            data = Path(path).read_bytes()
            assert is_step_bytes(data)
        finally:
            os.unlink(path)

    def test_file_matches_bytes_output(self):
        j = load_magnetic_json("concentric_basic.json")
        with tempfile.NamedTemporaryFile(suffix=".step", delete=False) as f:
            path = f.name
        try:
            mvbpp.draw_magnetic(j, path)
            file_data  = Path(path).read_bytes()
            bytes_data = mvbpp.draw_magnetic_to_bytes(j)
            # Both paths should produce STEP of similar size (within 1%)
            ratio = abs(len(file_data) - len(bytes_data)) / max(len(file_data), len(bytes_data))
            assert ratio < 0.01, f"File vs bytes size differs by {ratio:.1%}"
        finally:
            os.unlink(path)


# ── draw parameters (no config object) ────────────────────────────────────────

class TestDrawParameters:

    def test_default_parameters_produce_output(self):
        j = load_magnetic_json("concentric_basic.json")
        result = mvbpp.draw_magnetic_to_bytes(j)
        assert is_step_bytes(result)

    def test_no_bobbin_produces_smaller_output(self):
        j = load_magnetic_json("concentric_basic.json")
        with_bobbin    = mvbpp.draw_magnetic_to_bytes(j, include_bobbin=True)
        without_bobbin = mvbpp.draw_magnetic_to_bytes(j, include_bobbin=False)
        assert len(with_bobbin) > len(without_bobbin)

    def test_scale_affects_output(self):
        """Scale=1000 (mm) should be accepted without error."""
        j = load_magnetic_json("concentric_basic.json")
        result = mvbpp.draw_magnetic_to_bytes(j, scale=1000.0)
        assert is_step_bytes(result)

    def test_wire_polygon_segments_accepted(self):
        j = load_magnetic_json("concentric_basic.json")
        result = mvbpp.draw_magnetic_to_bytes(j, wire_polygon_segments=8)
        assert is_step_bytes(result)

    def test_core_polygon_segments_accepted(self):
        j = load_magnetic_json("concentric_basic.json")
        result = mvbpp.draw_magnetic_to_bytes(j, core_polygon_segments=32)
        assert is_step_bytes(result)

    def test_no_draw_config_class(self):
        """DrawConfig must not be exposed — configuration goes via kwargs only."""
        assert not hasattr(mvbpp, "DrawConfig")


# ── 2D dimensioned drawings ───────────────────────────────────────────────────

class TestDimensionedDrawings:

    def test_front_view_returns_svg(self):
        j = load_magnetic_json("concentric_basic.json")
        svg = mvbpp.draw_dimensioned_front_view(j)
        assert isinstance(svg, str)
        assert svg.lstrip().startswith("<?xml") or svg.lstrip().startswith("<svg")

    def test_top_view_returns_svg(self):
        j = load_magnetic_json("concentric_basic.json")
        svg = mvbpp.draw_dimensioned_top_view(j)
        assert isinstance(svg, str)
        assert svg.lstrip().startswith("<?xml") or svg.lstrip().startswith("<svg")

    def test_front_view_custom_colors(self):
        j = load_magnetic_json("concentric_basic.json")
        svg = mvbpp.draw_dimensioned_front_view(
            j, width_px=600.0, label_font_px=10.0,
            projection_color="#222222", dimension_color="#ff0000")
        assert "#ff0000" in svg or "ff0000" in svg.lower()

    def test_write_front_view_to_file(self):
        j = load_magnetic_json("concentric_basic.json")
        with tempfile.NamedTemporaryFile(suffix=".svg", delete=False) as f:
            path = f.name
        try:
            mvbpp.write_dimensioned_front_view(j, path)
            data = Path(path).read_text()
            assert "<svg" in data
        finally:
            os.unlink(path)


# ── symmetry ──────────────────────────────────────────────────────────────────

class TestSymmetry:

    def test_get_symmetry_planes_returns_list(self):
        j = load_magnetic_json("concentric_basic.json")
        planes = mvbpp.get_symmetry_planes(j)
        assert isinstance(planes, list)

    def test_e_core_has_symmetry_planes(self):
        j = load_magnetic_json("concentric_basic.json")
        planes = mvbpp.get_symmetry_planes(j)
        assert len(planes) >= 1, f"Expected at least 1 symmetry plane, got {planes}"

    def test_symmetry_plane_names_are_valid(self):
        j = load_magnetic_json("concentric_basic.json")
        planes = mvbpp.get_symmetry_planes(j)
        valid = {"XY", "YZ", "XZ"}
        for p in planes:
            assert p in valid, f"Unexpected plane name: {p}"

    def test_one_plane_produces_smaller_step(self):
        j = load_magnetic_json("concentric_basic.json")
        full = mvbpp.draw_magnetic_to_bytes(j, include_bobbin=False, symmetry_planes=0)
        half = mvbpp.draw_magnetic_to_bytes(j, include_bobbin=False, symmetry_planes=1)
        assert len(half) < len(full), "Half-symmetry STEP should be smaller than full"

    def test_two_planes_smaller_than_one_plane(self):
        j = load_magnetic_json("concentric_basic.json")
        planes = mvbpp.get_symmetry_planes(j)
        if len(planes) < 2:
            pytest.skip("Magnetic has fewer than 2 symmetry planes")
        half    = mvbpp.draw_magnetic_to_bytes(j, include_bobbin=False, symmetry_planes=1)
        quarter = mvbpp.draw_magnetic_to_bytes(j, include_bobbin=False, symmetry_planes=2)
        assert len(quarter) < len(half), "Quarter-symmetry STEP should be smaller than half"

    def test_symmetry_output_is_valid_step(self):
        j = load_magnetic_json("concentric_basic.json")
        result = mvbpp.draw_magnetic_to_bytes(j, include_bobbin=False, symmetry_planes=1)
        assert is_step_bytes(result)


# ── error handling ────────────────────────────────────────────────────────────

class TestErrorHandling:

    def test_invalid_json_raises(self):
        with pytest.raises(Exception):
            mvbpp.draw_magnetic_to_bytes("not json at all")

    def test_empty_json_object_raises(self):
        with pytest.raises(Exception):
            mvbpp.draw_magnetic_to_bytes("{}")

File: bindings/wasm/tests/test_mvbpp.js
/**
 * mvbpp WASM binding tests.
 *
 * Run with:  node bindings/wasm/tests/test_mvbpp.js
 * Requires:  bindings/wasm/dist/mvbpp.js  (built via `npm run build`)
 */

'use strict';

const fs   = require('fs');
const path = require('path');

const DIST_JS   = path.resolve(__dirname, '../dist/mvbpp.js');
const TESTS_DIR = __dirname;

// ── tiny test harness ─────────────────────────────────────────────────────────

let passed = 0, failed = 0;
const results = [];

function test(name, fn) {
    try {
        fn();
        console.log(`  ✓ ${name}`);
        passed++;
    } catch (e) {
        console.error(`  ✗ ${name}`);
        console.error(`    ${e.message}`);
        failed++;
    }
}

function assert(cond, msg) {
    if (!cond) throw new Error(msg || 'Assertion failed');
}

function assertEqual(a, b, msg) {
    if (a !== b) throw new Error(msg || `Expected ${b}, got ${a}`);
}

// ── helpers ───────────────────────────────────────────────────────────────────

function loadMagneticJson(filename) {
    const raw  = fs.readFileSync(path.join(TESTS_DIR, filename), 'utf8');
    const data = JSON.parse(raw);
    return JSON.stringify(data.magnetic);
}

function isStepBytes(buf) {
    // STEP files begin with "ISO-10303-21"
    const header = Buffer.from(buf.slice(0, 12));
    return header.toString('ascii') === 'ISO-10303-21';
}

function bufferToString(buf) {
    return Buffer.from(buf).toString('latin1');
}

// ── main ──────────────────────────────────────────────────────────────────────

async function main() {
    if (!fs.existsSync(DIST_JS)) {
        console.error(`ERROR: ${DIST_JS} not found. Run 'npm run build' in bindings/wasm/ first.`);
        process.exit(1);
    }

    const createMvbpp = require(DIST_JS);
    const mvbpp = await createMvbpp();

    console.log('\nmvbpp WASM — drawMagneticToBuffer');
    console.log('─'.repeat(50));

    // ── drawMagneticToBuffer ─────────────────────────────────────────────────

    test('returns Uint8Array for concentric_basic', () => {
        const j      = loadMagneticJson('concentric_basic.json');
        const result = mvbpp.drawMagneticToBuffer(j);
        assert(result instanceof Uint8Array, 'Expected Uint8Array');
    });

    test('output is non-empty', () => {
        const j      = loadMagneticJson('concentric_basic.json');
        const result = mvbpp.drawMagneticToBuffer(j);
        assert(result.length > 1000, `Too small: ${result.length} bytes`);
    });

    test('output has valid STEP header', () => {
        const j      = loadMagneticJson('concentric_basic.json');
        const result = mvbpp.drawMagneticToBuffer(j);
        assert(isStepBytes(result), 'Does not start with ISO-10303-21');
    });

    test('STEP contains solid geometry markers', () => {
        const j    = loadMagneticJson('concentric_basic.json');
        const text = bufferToString(mvbpp.drawMagneticToBuffer(j));
        assert(
            text.includes('CLOSED_SHELL') || text.includes('ADVANCED_BREP'),
            'No solid geometry markers found in STEP output'
        );
    });

    test('ETD49 produces valid STEP', () => {
        const j      = loadMagneticJson('ETD49_N87_10uH_5T.json');
        const result = mvbpp.drawMagneticToBuffer(j);
        assert(isStepBytes(result), 'ETD49 STEP header invalid');
        assert(result.length > 10_000, `ETD49 output too small: ${result.length}`);
    });

    test('ETD49 produces more geometry than concentric_basic', () => {
        const basic = mvbpp.drawMagneticToBuffer(loadMagneticJson('concentric_basic.json'));
        const etd   = mvbpp.drawMagneticToBuffer(loadMagneticJson('ETD49_N87_10uH_5T.json'));
        assert(etd.length > basic.length,
            `ETD49 (${etd.length}) should be larger than basic (${basic.length})`);
    });

    // ── drawMagnetic (to virtual FS path) ────────────────────────────────────

    console.log('\nmvbpp WASM — drawMagnetic (virtual FS)');
    console.log('─'.repeat(50));

    test('writes file to emscripten virtual FS', () => {
        const j    = loadMagneticJson('concentric_basic.json');
        const out  = '/tmp/test_basic.step';
        mvbpp.drawMagnetic(j, out);
        const data = mvbpp.FS.readFile(out);
        assert(data.length > 1000, `Virtual FS file too small: ${data.length}`);
        mvbpp.FS.unlink(out);
    });

    test('virtual FS file has valid STEP header', () => {
        const j   = loadMagneticJson('concentric_basic.json');
        const out = '/tmp/test_header.step';
        mvbpp.drawMagnetic(j, out);
        const data = mvbpp.FS.readFile(out);
        assert(isStepBytes(data), 'Virtual FS file does not start with ISO-10303-21');
        mvbpp.FS.unlink(out);
    });

    test('virtual FS file matches buffer output', () => {
        const j   = loadMagneticJson('concentric_basic.json');
        const out = '/tmp/test_match.step';
        mvbpp.drawMagnetic(j, out);
        const fileData   = mvbpp.FS.readFile(out);
        const bufferData = mvbpp.drawMagneticToBuffer(j);
        mvbpp.FS.unlink(out);
        const ratio = Math.abs(fileData.length - bufferData.length) /
                      Math.max(fileData.length, bufferData.length);
        assert(ratio < 0.01, `File vs buffer size differs by ${(ratio * 100).toFixed(1)}%`);
    });

    // ── DrawConfig ────────────────────────────────────────────────────────────

    console.log('\nmvbpp WASM — DrawConfig');
    console.log('─'.repeat(50));

    test('DrawConfig default produces valid STEP', () => {
        const j   = loadMagneticJson('concentric_basic.json');
        const cfg = new mvbpp.DrawConfig();
        const res = mvbpp.drawMagneticToBufferWithConfig(j, cfg);
        assert(isStepBytes(res), 'DrawConfig default: invalid STEP header');
        cfg.delete();
    });

    test('no bobbin produces smaller output', () => {
        const j = loadMagneticJson('concentric_basic.json');

        const cfgWith = new mvbpp.DrawConfig();
        cfgWith.includeBobbin = true;
        const withBobbin = mvbpp.drawMagneticToBufferWithConfig(j, cfgWith);
        cfgWith.delete();

        const cfgWithout = new mvbpp.DrawConfig();
        cfgWithout.includeBobbin = false;
        const withoutBobbin = mvbpp.drawMagneticToBufferWithConfig(j, cfgWithout);
        cfgWithout.delete();

        assert(withBobbin.length > withoutBobbin.length,
            `With bobbin (${withBobbin.length}) should be larger than without (${withoutBobbin.length})`);
    });

    test('scale=1000 accepted without error', () => {
        const j   = loadMagneticJson('concentric_basic.json');
        const cfg = new mvbpp.DrawConfig();
        cfg.scale = 1000.0;
        const res = mvbpp.drawMagneticToBufferWithConfig(j, cfg);
        assert(isStepBytes(res), 'scale=1000 produced invalid STEP');
        cfg.delete();
    });

    // ── error handling ────────────────────────────────────────────────────────

    console.log('\nmvbpp WASM — error handling');
    console.log('─'.repeat(50));

    test('invalid JSON throws', () => {
        let threw = false;
        try { mvbpp.drawMagneticToBuffer('not json'); } catch { threw = true; }
        assert(threw, 'Expected exception for invalid JSON');
    });

    test('empty object throws', () => {
        let threw = false;
        try { mvbpp.drawMagneticToBuffer('{}'); } catch { threw = true; }
        assert(threw, 'Expected exception for empty object');
    });

    // ── summary ───────────────────────────────────────────────────────────────

    console.log('\n' + '─'.repeat(50));
    console.log(`${passed + failed} tests: ${passed} passed, ${failed} failed`);

    if (failed > 0) process.exit(1);
}

main().catch(e => { console.error(e); process.exit(1); });
